### PR TITLE
Add better typing for CloudFunctions

### PR DESCRIPTION
**Pull request by inlined**
Trying to fork off a few smaller changes from my mondo refactor. We currently use the type of TriggerAnnotation as our type for CloudFunctions' API even though they don't line up 1:1. For example, there's the difference between failurePolicy vs eventType.failurePolicy and serviceAccount vs serviceAccountEmail.

This change introduces a new API type to protect us against type mismatches between trigger discovery and API calls.

### CODE REVIEW

File: src/gcp/cloudfunctions.ts [hunk 1]
```diff
@@ -20,17 +33,126 @@ interface Operation {
...
+export type CloudFunctionStatus =
+  | "ACTIVE"
+  | "OFFLINE"
+  | "DEPLOY_IN_PROGRESS"
+  | "DELETE_IN_PROGRESS"
+  | "UNKNOWN";
+export type SecurityLevel = "SECURE_ALWAYS" | "SECURE_OPTIONAL";
+
+export interface FailurePolicy {
+  // oneof action
```
Comment (joehan) [inline src/gcp/cloudfunctions.ts:83]: I see that you're still holding out hope for more options here 😄

File: src/gcp/proto.ts [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+        ","
+      )}`
+    );
+  }
+}
+
+// Utility function to help copy fields from type A to B.
+// As a safety net, catches typos or fields that aren't named the same
+// in A an B, but cannot verify that both Src and Dest have the same type for the same field.
+export function copyIfPresent<Src, Dest>(
```
Comment (joehan) [inline src/gcp/proto.ts:35]: Optional: Would be reasonable to add some quick unit tests for these helper functions. They're simple enough that its not blocking tho.

File: src/gcp/proto.ts [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+    throw new FirebaseError(
+      `Invalid ${typename} definition. ${oneof} can only have one field defined, but found ${defined.join(
+        ","
+      )}`
+    );
+  }
+}
+
+// Utility function to help copy fields from type A to B.
+// As a safety net, catches typos or fields that aren't named the same
```
Comment (joehan) [inline src/gcp/proto.ts:33]: > catches typos or fields that aren't named the same

Does it? it seems like this just silently does nothing in that case

File: src/gcp/cloudfunctions.ts [hunk 4]
```diff
@@ -20,17 +33,126 @@ interface Operation {
...
+  corsPolicy?: CorsPolicy;
+  vpcConnector?: string;
+  vpcConnectorEgressSettings?: "PRIVATE_RANGES_ONLY" | "ALL_TRAFFIC";
+  ingressSettings?: "ALLOW_ALL" | "ALLOW_INTERNAL_ONLY" | "ALLOW_INTERNAL_AND_GCLB";
+
+  kmsKeyName?: string;
+  buildWorkerPool?: string;
+  secretEnvironmentVariables?: SecretEnvVar[];
+  secretVolumes?: SecretVolume[];
+  sourceToken?: string;
```
Comment (joehan) [inline src/gcp/cloudfunctions.ts:134]: Should we add a comment that `sourceToken` is input only?

**Review by joehan — APPROVED** [review]
A few comments, but this seems pretty reasonable to me

File: src/gcp/cloudfunctions.ts [hunk 5]
```diff
@@ -5,7 +5,20 @@ import * as api from "../api";
 import { FirebaseError } from "../error";
 import { logger } from "../logger";
 import * as utils from "../utils";
-import { CloudFunctionTrigger } from "../deploy/functions/deploymentPlanner";
+import { Runtime } from "inspector";
```
Comment (mbleigh) [inline src/gcp/cloudfunctions.ts:8]: Is this intentional or was it an accidental auto-import? It seems to be shadowed below.

File: src/gcp/cloudfunctions.ts [hunk 6]
```diff
@@ -20,17 +33,126 @@ interface Operation {
   error?: { code: number; message: string };
 }
 
-export const API_VERSION = "v1";
+export interface HttpsTrigger {
+  // output only
+  url?: string;
```
Comment (mbleigh) [inline src/gcp/cloudfunctions.ts:38]: Could use `readonly` for this if we want to get super nitty ([docs](https://www.typescriptlang.org/docs/handbook/classes.html#readonly-modifier)).

File: src/gcp/cloudfunctions.ts (hunk 4 above)
Comment (mbleigh) [reply src/gcp/cloudfunctions.ts:134]: It's not though, is it? Couldn't you read this out of an existing function to pass to another?

File: src/gcp/proto.ts (hunk 2 above)
Comment (mbleigh) [reply src/gcp/proto.ts:35]: +1 if we're refactoring to improve code quality may as well improve test coverage.

File: src/test/deploy/functions/prompts.spec.ts [hunk 7]
```diff
@@ -87,11 +98,12 @@ describe("promptForFailurePolicies", () => {
   });
 
   it("should propmt if an existing function adds a failure policy", async () => {
-    const func: CloudFunctionTrigger = {
+    const func = {
       name: "projects/a/locations/b/functions/c",
       entryPoint: "",
       labels: {},
       environmentVariables: {},
+      runtime: "nodejs14" as gcf.Runtime,
```
Comment (mbleigh) [inline src/test/deploy/functions/prompts.spec.ts:106]: nit: Is this cast required? If so, I think I might prefer a plain string union type to a separately declared type. Having to cast string literals is kinda annoying.

File: src/gcp/cloudfunctions.ts (hunk 4 above)
Comment (joehan) [reply src/gcp/cloudfunctions.ts:134]: Strangely enough, its not returned from the API. I just double checked by adding some console logs to the functions deployment code, and sourceToken is never returned on a function by the API. For our deployment logic, we get it off of the operation metadata.

File: src/gcp/cloudfunctions.ts (hunk 1 above)
Comment (inlined) [reply src/gcp/cloudfunctions.ts:83]: I later defined a `FunctionSpec` type and just simplify it to `retry` but I want to document the proto accurately in the GCP folder.

File: src/gcp/proto.ts (hunk 3 above)
Comment (inlined) [reply src/gcp/proto.ts:33]: Yup. It's turned typos into compiler errors for me already. `keyof` is magical.

File: src/gcp/cloudfunctions.ts (hunk 5 above)
Comment (inlined) [reply src/gcp/cloudfunctions.ts:8]: Accidental auto-import

File: src/gcp/cloudfunctions.ts (hunk 6 above)
Comment (inlined) [reply src/gcp/cloudfunctions.ts:38]: SGTM

File: src/test/deploy/functions/prompts.spec.ts (hunk 7 above)
Comment (inlined) [reply src/test/deploy/functions/prompts.spec.ts:106]: This is a case of type forgetting. If we called this a gcf.CloudFunction then the literal string would compile completely OK, but we're currently using one literal to be both an `Omit<gcf.CloudFunction, gcf.OutputOnlyFields>` and as a `CloudFunctionTrigger` and the literal type just rembers "nodejs14" a string.

The final version of this code uses a `backend.FunctionSpec` for both input and output functions and the literal "nodejs14" is validated and remembered to be a `backend.Runtime` type.

**Review by mbleigh — APPROVED** [review]

**Pull request merged**